Add Aead Trait

l1h3r opened the pull request:
# Description of change

Adds an `Aead` trait to unify `AES-GCM` and `ChaCha` APIs (It will also be used for `AES-CBC` here: https://github.com/iotaledger/crypto.rs/pull/41).

Exports the following types to use with the trait:

- `Aes256Gcm`
- `XChaCha20Poly1305`

## Links to any relevant issues

Closes https://github.com/iotaledger/crypto.rs/issues/43

## Type of change

- [x] Enhancement (a non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

Ran existing tests and added additional test vectors for `Aes256Gcm`

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

nothingismagick reviewed (changes requested):
Since this is a breaking change, we would do well to document exactly how the new usage is different. At the very least we should put some documentation in the ciphers/traits.rs file

Otherwise this is a very forward thinking PR. :)

thibault-martinez commented on src/ciphers/traits.rs:
```diff
@@ -0,0 +1,97 @@
+        aad: &[u8],
+        ptx: &[u8],
+        ctx: &mut [u8],
+        tag: &mut Tag<Self::TagLength>,
```
This is probably just my lack of cipher knowledge but will all these parameters always make sense for all kind of ciphers ?

Wollac replied:
I agree, this is an interface for GCM, not even AEAD, or even [AE](https://en.wikipedia.org/wiki/Authenticated_encryption) yet alone general encryption i.e. cypher.
In that sense, does it rely provide the sense of abstraction that we are aiming for?

Wollac replied:
The very least, this should be called an AEAD trait. And use better names/descriptions for the parameters. Also nonce, and iv (initial vector) is not the same thing, so we should be consistent here.
Note that, `iv` and `aad` (which, I suppose, stands for associated data?) are optional, there are authenticating encryption schemes that do not rely on this.

Depending on the expected use-case, I would maybe propose to make this a GCM only trait and if necessary add a more general AE trait.

l1h3r replied:
Agreed that `Cipher` is not a good name but this is still an interface for AEAD, right? It was based on [`AeadInPlace`](https://docs.rs/aead/0.3.2/aead/trait.AeadInPlace.html) from the `aead` crate. I can go with the name `Aead`.

Will consistently use `nonce` and add more descriptive names 👍

Wollac reviewed (commented):
In contrast to the title, this PR does various things:
- it adds new "ciphers" e.g. chacha20poly1305
- it moves tests around and
- it introduces an AEAD trait for GCM and Poly1305

It don't think these should be in a single PR as they have different aspects to it.

What is the driving force behind this PR:
- Do we have a project that relies on easily exchanging of CHACHA-POLY1305 and AES-GCM?
- Do we need authenticated encryption *with associated data* in this context? This is required, e.g. for packets that have to contain a public header and encrypted payload, but not for "general" encryption of private data?
- Do we need the option to explicitly provide a nonce or could this be replaced by a random generator?
- Should the trait be compatible with other AE(AD) schemes?

l1h3r commented:
> In contrast to the title, this PR does various things:
> 
>     * it adds new "ciphers" e.g. chacha20poly1305
>     * it moves tests around and
>     * it introduces an AEAD trait for GCM and Poly1305
> 
> It don't think these should be in a single PR as they have different aspects to it.

Yeah, this PR is doing quite a bit. I will split it up into easier-to-consume PRs (tests/traits/impls)

> What is the driving force behind this PR:
> 
>     * Do we have a project that relies on easily exchanging of CHACHA-POLY1305 and AES-GCM?

My original motivation was to make things easier while working on JSON Web Tokens for identity.rs. Some examples of how it's being used are here:
https://github.com/iotaledger/identity.rs/blob/feat/libjose/libjose/src/jwe/encoder.rs#L342
https://github.com/iotaledger/identity.rs/blob/feat/libjose/libjose/src/jwe/encoder.rs#L628

I was planning to add AES-CBC which would also implement this trait:
https://github.com/iotaledger/crypto.rs/blob/libjose/src/aes_cbc.rs

>     * Do we need authenticated encryption _with associated data_ in this context? This is required, e.g. for packets that have to contain a public header and encrypted payload, but not for "general" encryption of private data?

Yeah, it will be used for identity-related communication - [DIDComm](https://identity.foundation/didcomm-messaging/spec/)

>     * Do we need the option to explicitly provide a nonce or could this be replaced by a random generator?

Generating one is fine for my needs but I'm not sure if that makes it too limiting for others

>     * Should the trait be compatible with other AE(AD) schemes?

Yes (to some extent I suppose). Personally I just want to add AES-CBC; I'm not sure what others would be needed

l1h3r committed:
Add Aead trait and implementations

l1h3r committed:
Add changelog message

l1h3r commented:
Reduced this PR down to just the addition of the new trait:

- [Separated the test structures changes](https://github.com/iotaledger/crypto.rs/pull/63) (this PR is still based on that branch)
- Removed additional algorithms - will restore will later PRs
- Renamed `Cipher` -> `Aead`
- Added `Aead::random_nonce`. Enabled with the current `random` feature
- Consistently used `nonce`; overall naming improvements
- Docs and doctests

l1h3r committed:
Add migration note

nothingismagick reviewed (approved):
OK, thanks for the revision and the migration notes.

Wollac reviewed (approved):
Thanks! It's much clearer now!
